File: KryossApi/src/KryossApi/Data/Entities/CopilotReadiness.cs
namespace KryossApi.Data.Entities;

public class CopilotReadinessScan
{
    public Guid Id { get; set; }
    public Guid OrganizationId { get; set; }
    public Guid TenantId { get; set; }
    public string Status { get; set; } = "running";
    public decimal? D1Score { get; set; }
    public decimal? D2Score { get; set; }
    public decimal? D3Score { get; set; }
    public decimal? D4Score { get; set; }
    public decimal? D5Score { get; set; }
    public decimal? D6Score { get; set; }
    public decimal? OverallScore { get; set; }
    public string? Verdict { get; set; }
    public string? PipelineStatus { get; set; }
    public DateTime StartedAt { get; set; }
    public DateTime? CompletedAt { get; set; }
    public DateTime CreatedAt { get; set; }

    public Organization Organization { get; set; } = null!;
    public M365Tenant Tenant { get; set; } = null!;
    public ICollection<CopilotReadinessMetric> Metrics { get; set; } = [];
    public ICollection<CopilotReadinessFinding> Findings { get; set; } = [];
    public ICollection<CopilotReadinessSharepoint> SharepointSites { get; set; } = [];
    public ICollection<CopilotReadinessExternalUser> ExternalUsers { get; set; } = [];
}

public class CopilotReadinessMetric
{
    public long Id { get; set; }
    public Guid ScanId { get; set; }
    public string Dimension { get; set; } = null!;
    public string MetricKey { get; set; } = null!;
    public string MetricValue { get; set; } = null!;
    public DateTime CreatedAt { get; set; }

    public CopilotReadinessScan Scan { get; set; } = null!;
}

public class CopilotReadinessFinding
{
    public long Id { get; set; }
    public Guid ScanId { get; set; }
    public string Service { get; set; } = null!;
    public string Feature { get; set; } = null!;
    public string Status { get; set; } = null!;
    public string Priority { get; set; } = "";
    public string? Observation { get; set; }
    public string? Recommendation { get; set; }
    public string? LinkText { get; set; }
    public string? LinkUrl { get; set; }
    public DateTime CreatedAt { get; set; }

    public CopilotReadinessScan Scan { get; set; } = null!;
}

public class CopilotReadinessSharepoint
{
    public long Id { get; set; }
    public Guid ScanId { get; set; }
    public string SiteUrl { get; set; } = null!;
    public string? SiteTitle { get; set; }
    public int TotalFiles { get; set; }
    public int LabeledFiles { get; set; }
    public int OversharedFiles { get; set; }
    public string? RiskLevel { get; set; }
    public string? TopLabels { get; set; }
    public DateTime CreatedAt { get; set; }

    public CopilotReadinessScan Scan { get; set; } = null!;
}

public class CopilotReadinessExternalUser
{
    public long Id { get; set; }
    public Guid ScanId { get; set; }
    public string UserPrincipal { get; set; } = null!;
    public string? DisplayName { get; set; }
    public string? EmailDomain { get; set; }
    public DateTime? LastSignIn { get; set; }
    public string? RiskLevel { get; set; }
    public int SitesAccessed { get; set; }
    public string? HighestPermission { get; set; }
    public DateTime CreatedAt { get; set; }

    public CopilotReadinessScan Scan { get; set; } = null!;
}

File: KryossPortal/src/api/copilotReadiness.ts
import { useQuery, useMutation, useQueryClient } from '@tanstack/react-query';
import { apiFetch } from './client';

// ── Types ──

export interface CopilotReadinessScan {
  id: string;
  status: 'running' | 'completed' | 'partial' | 'failed';
  d1Score: number | null;
  d2Score: number | null;
  d3Score: number | null;
  d4Score: number | null;
  d5Score: number | null;
  d6Score: number | null;
  overallScore: number | null;
  verdict: string | null;
  pipelineStatus: Record<string, string>;
  startedAt: string;
  completedAt: string | null;
  findingsSummary: Record<string, { total: number; actionRequired: number; warning: number; success: number }>;
}

export interface CopilotReadinessFinding {
  id: number;
  service: string;
  feature: string;
  status: string;
  priority: string;
  observation: string | null;
  recommendation: string | null;
  linkText: string | null;
  linkUrl: string | null;
}

export interface CopilotReadinessScanDetail extends CopilotReadinessScan {
  findings: CopilotReadinessFinding[];
  metrics: Record<string, string>;
  sharepointSites: Array<{
    siteUrl: string;
    siteTitle: string;
    totalFiles: number;
    labeledFiles: number;
    oversharedFiles: number;
    riskLevel: string;
  }>;
  externalUsers: Array<{
    userPrincipal: string;
    displayName: string;
    emailDomain: string;
    lastSignIn: string | null;
    riskLevel: string;
    sitesAccessed: number;
  }>;
}

export interface CopilotReadinessScanSummary {
  id: string;
  overallScore: number;
  verdict: string;
  createdAt: string;
}

// ── Hooks ──

export function useCopilotReadiness(organizationId: string | undefined) {
  return useQuery({
    queryKey: ['copilot-readiness', organizationId],
    queryFn: () =>
      apiFetch<CopilotReadinessScan | { scanned: false }>(
        `/v2/copilot-readiness?organizationId=${organizationId}`,
      ),
    enabled: !!organizationId,
    refetchInterval: (query) => {
      const data = query.state.data;
      return data && 'status' in data && data.status === 'running' ? 10000 : false;
    },
  });
}

export function useCopilotReadinessDetail(scanId: string | undefined) {
  return useQuery({
    queryKey: ['copilot-readiness-detail', scanId],
    queryFn: () =>
      apiFetch<CopilotReadinessScanDetail>(`/v2/copilot-readiness/${scanId}`),
    enabled: !!scanId,
  });
}

export function useCopilotReadinessHistory(organizationId: string | undefined) {
  return useQuery({
    queryKey: ['copilot-readiness-history', organizationId],
    queryFn: () =>
      apiFetch<CopilotReadinessScanSummary[]>(
        `/v2/copilot-readiness/history?organizationId=${organizationId}`,
      ),
    enabled: !!organizationId,
  });
}

export function useCopilotReadinessScan() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: (organizationId: string) =>
      apiFetch<{ scanId: string; status: string }>('/v2/copilot-readiness/scan', {
        method: 'POST',
        body: JSON.stringify({ organizationId }),
      }),
    onSuccess: (_data, organizationId) => {
      qc.invalidateQueries({ queryKey: ['copilot-readiness', organizationId] });
    },
  });
}
